### PR TITLE
Add helm flags to control the app probe rewrite beahvior.

### PR DESCRIPTION
**Pull request by incfly**
For https://github.com/istio/istio/issues/9150

`istioctl kube-inject` and auto injection has slightly different code path. The existing override changes only happen on istioctl level.

`.Values.global.sidecarInjectorWebhook.rewriteAppHTTPProbe` controls the sidecar injection template's option `rewriteAppHTTPProbe`. During injection, we'll then consult this flag to determine whether spec will be rewritten.

### CODE REVIEW

File: istioctl/cmd/istioctl/kubeinject.go [hunk 1]
```diff
@@ -274,6 +274,7 @@ istioctl kube-inject -f deployment.yaml -o deployment-injected.yaml --injectConf
 				if sidecarTemplate, err = inject.GenerateTemplateFromParams(&inject.Params{
 					InitImage:                    inject.InitImageName(hub, tag, debugMode),
 					ProxyImage:                   inject.ProxyImageName(hub, tag, debugMode),
+					RewriteAppHTTPProbe:          false,
```
Comment (nmittler) [inline istioctl/cmd/istioctl/kubeinject.go]: should this be a flag?

File: pilot/pkg/kube/inject/webhook.go [hunk 2]
```diff
@@ -483,6 +483,7 @@ func (wh *Webhook) inject(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionRespons
 		}
 	}
 
+	// TODO: plumbing wh.sidecarConfig.RewriteHTTPPRobe to the injectionData.
```
Comment (nmittler) [inline pilot/pkg/kube/inject/webhook.go]: nit: add your github handle in the TODO

Comment (incfly) [reply pilot/pkg/kube/inject/webhook.go]: Done.

File: istioctl/cmd/istioctl/kubeinject.go (hunk 1 above)
Comment (incfly) [reply istioctl/cmd/istioctl/kubeinject.go]: Done.

**Comment by tiswanso** [comment]
This PR is causing e2e-simple to fail when run with auto-sidecar injection.  Seems like the new helm chart param `sidecarInjectorWebhook.rewriteAppHTTPProbe` should default to `true` (maybe only when auth is enabled?).

The istio-cni release-1.1 nightly is failing due to it.

File: install/kubernetes/helm/istio/values.yaml [hunk 3]
```diff
@@ -11,6 +11,7 @@ gateways:
 #
 sidecarInjectorWebhook:
   enabled: true
+  rewriteAppHTTPProbe: false
```
[the remark this answers is not in the document]
Comment (costinm) [reply install/kubernetes/helm/istio/values.yaml]: Once again - after reverting please add comments explaining what it does. Nobody can understand - rewrite how, when to set, why, etc. Don't assume all users are familiar with the your implementation details and goals.

File: pilot/pkg/kube/inject/app_probe_test.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ func TestRewriteAppHTTPProbe(t *testing.T) {
 		{
 			name: "one-container",
 			sidecar: &SidecarInjectionSpec{
+				RewriteAppHTTPProbe: true,
```
Comment (costinm) [inline pilot/pkg/kube/inject/app_probe_test.go]: This doesn't make any sense - why would this be part of the spec, next to containers ?

**Comment by nmittler** [comment]
@incfly can you address the concerns above from @costinm?

**Comment by incfly** [comment]
- I'll first take a look at the breakage first. `rewriteAppHTTPProbe` is default to false to retain the current behavior. Not sure where it introduce the delta.
- It **is** possible to gradually rolled out. You don't turn on `rewriteAppProbe` and `istio kube-inject` has additional flag to control on per deployment base.
- I *do remember* to add an e2e test for this. Look at some issues, https://github.com/istio/istio/issues/10524. Right now the helm flag is off everywhere.

**Comment by tiswanso** [comment]
To add info on what I've seen with this change... I think this has introduced a delta (at least for e2e-simple) for the e2e test framework run with sidecar injection enabled v. the default with istioctl kube-inject.  I don't see how the e2e-simple would pass without the rewriteAppHTTPProbe=true so, I'm suspecting that the e2e test is running with it set for the istioctl kube-inject case.  IMO, that delta in behavior is really bad and should be fixed by either changing the default for the flag or by having the makefile set it all the time for the e2e tests.

**Comment by incfly** [comment]
I filed bug https://github.com/istio/istio/issues/10557. Let's move AI(s) over there.

**Comment by tiswanso** [comment]
@incfly -- I think pilot/pkg/kube/inject/app_probe.go L51-53 is what changed the default behavior--at least compared to master.  I'm not familiar with what was happening for the liveness probes prior to the existence app_probe.go.

**Comment by costinm** [comment]
kube-inject is not an option - many users use automatic sidecar injector, and it is what we recommend. 

Please use a pod annotation - so users can opt in and out per pod, like we do with many other features ( including mixer, policy, etc). There is not use case for having this as (only) a global setting.

**Pull request merged**